protocol/op-supervisor-integration.md

```diff
@@ -0,0 +1,309 @@
+    CrossUnsafe SafetyLevel = "cross-unsafe"
+    Unsafe      SafetyLevel = "unsafe"
+	Unknown     SafetyLevel = "unknown"
+    Conflicts   SafetyLevel = "conflicts"
+)
+
+type RPC interface {
+	CheckMessage(identifier types.Identifier,
+                 payloadHash common.Hash,
+                 ) (types.SafetyLevel, error)
+}
+```
+
+This method is used to verify initiating messages exist, e.g. used by the tx-pool and block-building, to verify user-input before inclusion into the chain.
+
+API implementation:
+
+1. Find the event log-db entry at the given identifier
+2. Reconstruct a `TruncatedHash` from the `identifier` and `payloadHash`
+3. Check if the log-db entry matches the reconstructed `TruncatedHash`
+    - If it matches, return the safety-level of the entry
+        - The safety is determined by the first (and thus strongest safety level) type of head of the chain that is past or equal the entry.
+    - If it does not match, return the `conflicts` safety level.
+    - If the entry is not yet known, return an `unknown` safety level.
+
+
+### `TryCrossUnsafe`
+
+```go
+type RPC interface {
+    GetCrossUnsafe(chainID types.ChainID, maxNumber uint64) (crossUnsafe eth.BlockID, err error)
+}
+```
+
+API implementation:
+1. Compare the `maxNumber` to the locally known `unsafe` tip of the event-DB:
+    - If newer:
+      1. Initiate fetching of new event-log data.
+      2. Return the latest known `cross-unsafe`.
+    - If equal or older:
+      1. Return the latest known `cross-unsafe` block before `maxNumber`.
+         This is the last block checkpoint `<= maxNumber`
+
+By clipping using `maxNumber` the op-node is able to verify the returned `cross-unsafe`, and not accept it if it's non-canonical, and optionally query for older cross-unsafe blocks to try to bump to.
+
+### `NextDeriveTask`, `OnDerived`, `CrossSafe`
+
+```go
+type RPC interface {
+	
+	// Returns where the next L2 block has to be derived from. 
+	NextDeriveTask(chainID types.ChainID) (parentL2 eth.BlockID, depositOnly bool, l1 eth.BlockID) error
+
+	// Signals what has been derived, considered to be local-safe in op-node.
+	OnDerived(chainID types.ChainID, derived eth.BlockID, parentL2 eth.BlockID, depositOnly bool, l1 eth.BlockID) error
+
+	// CrossSafe returns the cross-safe block, and where in L1 traversal it was recognized as cross-safe. 
+	CrossSafe(chainID types.ChainID, l1Number uint64) (crossSafe eth.BlockID, derivedFrom eth.BlockID, error)
+}
+```
+`NextDeriveTask`: Given the chainID, check if we have a local-safe block that builds on top of the latest cross-unsafe.
+- If we do not have such block, then create the task to derive one:
+  - `{parentL2: latestCrossSafe, depositOnly: false, l1: l1OfLatestCrossSafe}`
+- If we do, then determine if the local-safe block has been verified:
+  1. Check if all dependencies are:
+    - known
+    - cross-safe if previous height
+    - local-safe if same height
+    - have valid intra-block dependencies if same height
```

Comment:
What do we do if all dependencies are met? I'm guessing at that point we're waiting on other chains to complete there tasks before we can advance the cross-safe head. So this would return a no-op implementation and op-node should call again after a short delay.

Reply:
We progress the cross-safe head of a L2 (by inserting this cross-safe update as entry in the DB) if the dependencies are met. If the dependencies are non-existent or just older, then the L2 chain can become cross-safe before the other L2s become cross-safe. Not every L1 block may cause all L2s to progress their cross-safe at the same time.

An "unknown" error may still be returned, since checking those transitive dependencies, even if all input data is there, may not have completed yet. But if we know which L2 derived up to where from the L1 block at hand, and have the event-dependencies checked, then there is no need to delay.